File: src/main/kotlin/com/xs0/gqlktx/exec/FieldPath.java
package com.xs0.gqlktx.exec;

import io.vertx.core.json.JsonArray;

public class FieldPath {
    private final FieldPath parent;
    private final String fieldName;
    private final int listIndex;

    private FieldPath(FieldPath parent, String fieldName, int listIndex) {
        this.parent = parent;
        this.fieldName = fieldName;
        this.listIndex = listIndex;
    }

    public static FieldPath root() {
        return new FieldPath(null, null, 0);
    }

    public FieldPath subField(String fieldName) {
        return new FieldPath(this, fieldName, 0);
    }

    public FieldPath listElement(int listIndex) {
        return new FieldPath(this, null, listIndex);
    }

    public JsonArray toArray() {
        JsonArray result = new JsonArray();
        toArray(result);
        return result;
    }

    private void toArray(JsonArray result) {
        if (parent != null) {
            parent.toArray(result);
            if (fieldName != null) {
                result.add(fieldName);
            } else {
                result.add(listIndex);
            }
        }
    }

    @Override
    public String toString() {
        return toArray().toString();
    }

    public boolean isRoot() {
        return parent == null;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/exec/FieldPath.kt
package com.xs0.gqlktx.exec

class FieldPath private constructor(private val parent: FieldPath?, private val fieldName: String?, private val listIndex: Int?) {

    fun subField(fieldName: String): FieldPath {
        return FieldPath(this, fieldName, 0)
    }

    fun listElement(listIndex: Int): FieldPath {
        return FieldPath(this, null, listIndex)
    }

    fun toArray(): List<Any?> {
        val result = ArrayList<Any?>()
        toArray(result)
        return result
    }

    private fun toArray(result: ArrayList<Any?>) {
        if (parent != null) {
            parent.toArray(result)

            // not a bug - if parent is null, we are root and path is empty..
            if (fieldName != null) {
                result.add(fieldName)
            } else {
                result.add(listIndex)
            }
        }
    }

    override fun toString(): String {
        return toArray().toString()
    }

    val isRoot: Boolean
        get() = parent == null

    companion object {

        fun root(): FieldPath {
            return FieldPath(null, null, 0)
        }
    }
}
